source/module_hamilt_lcao/module_gint/gint_tools.cpp

```diff
@@ -97,64 +97,190 @@ void get_vindex(const int bxyz, const int bx, const int by, const int bz,
+	if (distance < 1.0E-9) { distance += 1.0E-9;
+}
+}
 
-			for(int ib=0; ib<bxyz; ib++)
-			{
-				double*const p_dpsi_x=&dpsir_ylm_x[ib][block_index[id]];
-				double*const p_dpsi_y=&dpsir_ylm_y[ib][block_index[id]];
-				double*const p_dpsi_z=&dpsir_ylm_z[ib][block_index[id]];
-				double*const p_dpsirr=&dpsirr_ylm[ib][block_index[id] * 6];
-				if(!cal_flag[ib][id])
-				{
-					ModuleBase::GlobalFunc::ZEROS(p_dpsirr, block_size[id] * 6);
-				}
-				else
-				{
-					const double dr[3]={						// vectors between atom and grid
-						gt.meshcell_pos[ib][0] + mt[0],
-						gt.meshcell_pos[ib][1] + mt[1],
-						gt.meshcell_pos[ib][2] + mt[2]};
 
-					for (int iw=0; iw< atom->nw; ++iw)
-					{
-						p_dpsirr[iw * 6] = p_dpsi_x[iw]*dr[0];
-						p_dpsirr[iw * 6 + 1] = p_dpsi_x[iw]*dr[1];
-						p_dpsirr[iw * 6 + 2] = p_dpsi_x[iw]*dr[2];
-						p_dpsirr[iw * 6 + 3] = p_dpsi_y[iw]*dr[1];
-						p_dpsirr[iw * 6 + 4] = p_dpsi_y[iw]*dr[2];
-						p_dpsirr[iw * 6 + 5] = p_dpsi_z[iw]*dr[2];
-					}//iw
-				}//else
-			}
+void spl_intrp(const double distance,
+							const double delta_r,
+							Atom*& atom,
+							std::vector<double>& ylma,
+							std::vector<const double*>& it_psi_uniform,
+							std::vector<const double*>& it_dpsi_uniform,
+							double *p)
+{
+	double coeffs[4] = {0.0};
+	const double position = distance / delta_r;
+	int ip = static_cast<int>(position);
+	const double dx = position - ip;
+	const double dx2 = dx * dx;
+	const double dx3 = dx2 * dx;
+	coeffs[2] = 3.0 * dx2 - 2.0 * dx3;
+	coeffs[0] = 1.0 - coeffs[2];
+	coeffs[1] = (dx - 2.0 * dx2 + dx3) * delta_r;
+	coeffs[3] = (dx3 - dx2) * delta_r;
+
+	double phi = 0;
+	for (int iw = 0; iw < atom->nw; ++iw)
+	{
+		if (atom->iw2_new[iw])
+		{
+			auto psi_uniform = it_psi_uniform[iw];
+			auto dpsi_uniform = it_dpsi_uniform[iw];
+			phi = coeffs[0] * psi_uniform[ip] + coeffs[1] * dpsi_uniform[ip] // radial wave functions
+					+ coeffs[2] * psi_uniform[ip + 1] + coeffs[3] * dpsi_uniform[ip + 1];
 		}
-		ModuleBase::timer::tick("Gint_Tools", "cal_dpsirr_ylm");
-		return;
-	}
+		p[iw] = phi * ylma[atom->iw2_ylm[iw]];
+	} // end iw
+}
+
+void dpsi_spl_intrp(const double distance,
+								const double* dr,
+								const double delta_r,
+								Atom*& atom,
+                                double* rly,
+								double** grly,
+                                std::vector<const double*>& it_psi_uniform,
+                                std::vector<const double*>& it_dpsi_uniform,
+                                double *p_psi,
+								double *p_dpsi_x,
+								double *p_dpsi_y,
+								double *p_dpsi_z)
+{
+	const double position = distance / delta_r;
+
+	const double iq = static_cast<int>(position);
+	const int ip = static_cast<int>(position);
+	const double x0 = position - iq;
+	const double x1 = 1.0 - x0;
+	const double x2 = 2.0 - x0;
+	const double x3 = 3.0 - x0;
+	const double x12 = x1 * x2 / 6;
+	const double x03 = x0 * x3 / 2;
+
+	double tmp, dtmp;
```

Comment:
one line for one variable, initialize the variables

Reply:
Modified to initialized.